flow/scripts/macro_place.tcl

```diff
@@ -41,14 +48,51 @@ proc find_macros {} {
 
 
 if {[find_macros] != ""} {
-  if {[info exists ::env(MACRO_PLACEMENT)]} {
-    source $::env(SCRIPTS_DIR)/read_macro_placement.tcl
-    puts "\[INFO\]\[FLOW-xxxx\] Using manual macro placement file $::env(MACRO_PLACEMENT)"
-    read_macro_placement $::env(MACRO_PLACEMENT)
+  if {[info exists ::env(RTLMP_FLOW)]} {
+    puts "RTLMP Flow enabled..."
+    partition_design -max_num_inst $env(RTLMP_MAX_INST) \
+                    -min_num_inst $env(RTLMP_MIN_INST) \
+                    -max_num_macro $env(RTLMP_MAX_MACRO) \
+                    -min_num_macro $env(RTLMP_MIN_MACRO) \
+                    -net_threshold 5 -virtual_weight 500 \
+                    -num_hop 5 -timing_weight 1 \
+                    -report_directory $env(RTLMP_RPT_DIR) \
+                    -report_file $env(RTLMP_RPT_FILE)
+ 
+    set flag [rtl_macro_placer -config_file $env(RTLMP_CONFIG_FILE) \
+                                -macro_blockage_file $env(RTLMP_BLOCKAGE_FILE) \
+                                -report_directory $env(RTLMP_RPT_DIR) \
+                                -report_file $env(RTLMP_RPT_FILE)]
+    puts "Delete buffers for RTLMP flow..."
+    remove_buffers
   } else {
-    macro_placement \
-      -halo $::env(MACRO_PLACE_HALO) \
-      -channel $::env(MACRO_PLACE_CHANNEL)
+    if {[info exists ::env(MACRO_PLACEMENT)]} {
+      source $::env(SCRIPTS_DIR)/read_macro_placement.tcl
+      puts "\[INFO\]\[FLOW-xxxx\] Using manual macro placement file $::env(MACRO_PLACEMENT)"
+      read_macro_placement $::env(MACRO_PLACEMENT)
+    } else {
+      macro_placement \
+        -halo $::env(MACRO_PLACE_HALO) \
+        -channel $::env(MACRO_PLACE_CHANNEL)
+    }
+  }
+
+# If wrappers defined replace macros with their wrapped version
+# # ----------------------------------------------------------------------------
+  if {[info exists ::env(MACRO_WRAPPERS)]} {
+    source $::env(MACRO_WRAPPERS)
+
+    set wrapped_macros [dict keys [dict get $wrapper around]]
+    set db [ord::get_db]
+    set block [ord::get_db_block]
+
+    foreach inst [$block getInsts] {
+      if {[lsearch -exact $wrapped_macros [[$inst getMaster] getName]] > -1} {
+        set new_master [dict get $wrapper around [[$inst getMaster] getName]]
+        puts "Replacing [[$inst getMaster] getName] with $new_master for [$inst getName]"
+        $inst swapMaster [$db findMaster $new_master]
+      }
+    }
```

Comment:
There is an issue with moving this block here. LEF wrapping changes the area and pin placements of the design. You are moving this to after the RTLMP placement. This could yield illegal or suboptimal placements. I'm ok if you want to move this from floorplan.tcl but it would need to be the first thing in this file.

Reply:
There is no placement before rtl-mp in that flow.  We need the unwrapped macros as that's what is referred to in the verilog.

Reply:
Ok I see. Well then it at least needs to go before `macro_placement`.

Reply:
Currently in the flow it has to be called after macro placement as the RTLMP macro placer reads the verilog netlist. Once it goes into OpenDB it is flattened. Once we integrate design_browser we should be able to streamline this flow further

Reply:
@ravi-varadarajan this is still a change in behavior for the non-RTLMP flow though. The macro wrapping needs to happen before the `macro_placement` command in the default flow otherwise the results will change and the channel widths will be shorter due to wrapping after macro placement.

Reply:
@ravi-varadarajan I think this will affect rtl-mp in the same way.  It seems like you would want to wrap after the netlist is loaded but before macro placement happens.

Reply:
In the flow when RTL-MP is run it  uses both the hierarchical netlist from verilog (openSTA) for hierarchy information and clustering  and OpenDB for the connectivity. Also the design read for macro_place.tcl is different or TMP flow which reads the db from the preivous step (tdms_place).  For RTL-MP if we wrap the macros before clustering and macro placement the OpenSTA datamodel will be out of sync with the wrapped memories.  Is that right?

Also wrapping after macro placement does seem to change the secure-ci results.  This should be cleaned up with the design_browser integration and reading DB.

Reply:
When you swap the master for a cell, the database issues a callback to all interested clients.  sta is register as a client and updates its data model (see dbStaCbk::inDbInstSwapMasterAfter).  There should be no issue replacing right before macro placement and that will solve the secure-ci issue.

Reply:
I did verify that it will be OK to move the macro wrapper part to before the macro placement as the hierarchy information that is captured is only the instance names (and not the masters)  I am running some tests locally and if they all pass I will push the change to move the wrapping before macro placement (RTLMP or TMP)

Reply:
Moved the macro wrapper code to before macro placements in macro_place.tcl